content/ansible/roles/manageiq-core.manageiq-vmdb/library/manageiq_vmdb.py

```diff
@@ -0,0 +1,199 @@
+#! /usr/bin/python
+
+from __future__ import (absolute_import, division, print_function)
+import os
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+module: manageiq_vmdb
+'''
+import json
+import re
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url
+from ansible.module_utils.six.moves.urllib.parse import urlparse
+
+class ManageIQVmdb(object):
+    """
+        Object to execute VMDB management operations in manageiq.
+    """
+
+    def __init__(self, module):
+        self._module = module
+        self._debug = bool(self._module._verbosity >= 3)
+        self._api_url = self._module.params['manageiq_connection']['url']
+        self._vmdb = self._module.params.get('vmdb') or self._module.params.get('href')
+        self._href = None
+        self._error = None
+        self._auth = self._build_auth()
+
+
+    def _build_auth(self):
+        self._headers = {'Content-Type': 'application/json; charset=utf-8'}
+        # Force CERT validation to work with fetch_url
+        self._module.params['validate_certs'] = self._module.params['manageiq_connection']['manageiq_validate_certs']
+        for cert in ('force_basic_auth', 'client_cert', 'client_key'):
+            self._module.params[cert] = self._module.params['manageiq_connection'][cert]
+        if self._module.params['manageiq_connection'].get('token'):
+            self._headers["X-Auth-Token"] = self._module.params['manageiq_connection']['token']
+        else:
+            self._module.params['url_username'] = self._module.params['manageiq_connection']['username']
+            self._module.params['url_password'] = self._module.params['manageiq_connection']['password']
+
+
+
+
+    @property
+    def url(self):
+        """
+            The url to connect to the VMDB Object
+        """
+        return self.build_url()
+
+
+    def build_url(self):
+        """
+            Using any type of href input, build out the correct url
+        """
+
+        url_actual = urlparse(self._href)
+        if re.search('api', url_actual.path):
+            return self._api_url + url_actual.path
+        return self._api_url + '/api/' + url_actual.path
```

Comment:
Minor but why have a separate url and build_url method?  I feel like the content of build_url could just be in url itself.

Reply:
@Fryguy - If I remember correctly - we don't have control over what comes in via the playbook.  There are cases where a slug is used `href_slug::users/10`  or other cases where `https://manageiq_domain.local/api/users/10` could be used. I didn't want the url 'building' logic to muddle up what I thought should be a `url` property in the class.
I think at the time, I thought the complexity would be worse than it ended up, so the logic in this case isn't too much of a deal, but just giving the background.